In [ ]:
#Import libraries
import pandas as pd
import numpy as np

In [ ]:
#import dataset
df = pd.read_csv('/content/drive/MyDrive/Superstore.csv', encoding='windows-1252')

In [ ]:
df.head()

Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
0       1  CA-2013-152156  09-11-2013  12-11-2013    Second Class    CG-12520   
1       2  CA-2013-152156  09-11-2013  12-11-2013    Second Class    CG-12520   
2       3  CA-2013-138688  13-06-2013  17-06-2013    Second Class    DV-13045   
3       4  US-2012-108966  11-10-2012  18-10-2012  Standard Class    SO-20335   
4       5  US-2012-108966  11-10-2012  18-10-2012  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
0       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1       42420   South  FUR-CH-10000454        Furniture       Chairs   
2       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3       33311   South  FUR-TA-10000577        Furniture       Tables   
4       33311   South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 21 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

##Data Cleaning

In [ ]:
#Ubah order date menjadi datetime
df['Order Date'] = pd.to_datetime(df['Order Date']).dt.date.astype('datetime64')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9994 non-null   int64         
 1   Order ID       9994 non-null   object        
 2   Order Date     9994 non-null   datetime64[ns]
 3   Ship Date      9994 non-null   object        
 4   Ship Mode      9994 non-null   object        
 5   Customer ID    9994 non-null   object        
 6   Customer Name  9994 non-null   object        
 7   Segment        9994 non-null   object        
 8   Country        9994 non-null   object        
 9   City           9994 non-null   object        
 10  State          9994 non-null   object        
 11  Postal Code    9994 non-null   int64         
 12  Region         9994 non-null   object        
 13  Product ID     9994 non-null   object        
 14  Category       9994 non-null   object        
 15  Sub-Category   9994 n

In [ ]:
#Cek null value
df.isna().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

In [ ]:
#cek duplicate
df.duplicated().sum()

0

In [ ]:
df.head()

Row ID        Order ID Order Date   Ship Date       Ship Mode Customer ID  \
0       1  CA-2013-152156 2013-09-11  12-11-2013    Second Class    CG-12520   
1       2  CA-2013-152156 2013-09-11  12-11-2013    Second Class    CG-12520   
2       3  CA-2013-138688 2013-06-13  17-06-2013    Second Class    DV-13045   
3       4  US-2012-108966 2012-11-10  18-10-2012  Standard Class    SO-20335   
4       5  US-2012-108966 2012-11-10  18-10-2012  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
0       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1       42420   South  FUR-CH-10000454        Furniture       Chairs   
2       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3       33311   South  FUR-TA-10000577        Furniture       Tables   
4       33311   South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 21 columns]

##RFM Analysis

Agregat data transaksi ke bentuk summary total transaksi(order count), total nilai order(order value), tanggal order terakhir dari setiap pengguna

In [ ]:
df_user = df.groupby('Customer ID', as_index=False).agg(
    order_count=('Order ID','nunique'),
    max_order_date=('Order Date', 'max'),
    total_order_value=('Sales', 'sum')
)

df_user

Customer ID  order_count max_order_date  total_order_value
0      AA-10315            5     2014-06-30           5563.560
1      AA-10375            9     2014-12-12           1056.390
2      AA-10480            4     2014-04-16           1790.512
3      AA-10645            6     2014-06-11           5086.935
4      AB-10015            3     2013-11-11            886.156
..          ...          ...            ...                ...
788    XP-21865           11     2014-11-18           2374.658
789    YC-21895            5     2014-12-27           5454.350
790    YS-21880            8     2014-12-22           6720.444
791    ZC-21910           13     2014-12-06           8025.707
792    ZD-21925            5     2014-12-06           1493.944

[793 rows x 4 columns]

Buat kolom jumlah hari sejak order terakhir

In [ ]:
#Diasumsikan hari terakhir order adalah tanggal maksmimum dari dataset
today = df['Order Date'].max()
df_user['day_since_last_order'] = (today - df_user['max_order_date']).dt.days
df_user

Customer ID  order_count max_order_date  total_order_value  \
0      AA-10315            5     2014-06-30           5563.560   
1      AA-10375            9     2014-12-12           1056.390   
2      AA-10480            4     2014-04-16           1790.512   
3      AA-10645            6     2014-06-11           5086.935   
4      AB-10015            3     2013-11-11            886.156   
..          ...          ...            ...                ...   
788    XP-21865           11     2014-11-18           2374.658   
789    YC-21895            5     2014-12-27           5454.350   
790    YS-21880            8     2014-12-22           6720.444   
791    ZC-21910           13     2014-12-06           8025.707   
792    ZD-21925            5     2014-12-06           1493.944   

     day_since_last_order  
0                     184  
1                      19  
2                     259  
3                     203  
4                     415  
..                    ...  
788                    43  
789                     4  
790                     9  
791                    25  
792                    25  

[793 rows x 5 columns]

In [ ]:
df_user.describe()

order_count  total_order_value  day_since_last_order
count   793.000000         793.000000            793.000000
mean      6.316520        2896.848500            163.928121
std       2.550885        2628.670117            189.632742
min       1.000000           4.833000              0.000000
25%       5.000000        1146.050000             37.000000
50%       6.000000        2256.394000             93.000000
75%       8.000000        3785.276000            226.000000
max      17.000000       25043.050000           1166.000000

Buat binning dari jumlah hari sejak order terakhir yang terdiri dari 5 bins dengan batas-batasnya merupakan min, P20, P40, P60, P80, max dan beri label 1 sampai 5 dari bin tertinggi ke terendah sebagai skor recency

In [ ]:
df_user['recency_score'] = pd.cut(df_user['day_since_last_order'],
                                  bins=[df_user['day_since_last_order'].min(),
                                        np.percentile(df_user['day_since_last_order'], 20),
                                        np.percentile(df_user['day_since_last_order'], 40),
                                        np.percentile(df_user['day_since_last_order'], 60),
                                        np.percentile(df_user['day_since_last_order'], 80),
                                        df_user['day_since_last_order'].max()],
                                  labels=[5, 4, 3, 2, 1],
                                  include_lowest=True).astype(int)
df_user

Customer ID  order_count max_order_date  total_order_value  \
0      AA-10315            5     2014-06-30           5563.560   
1      AA-10375            9     2014-12-12           1056.390   
2      AA-10480            4     2014-04-16           1790.512   
3      AA-10645            6     2014-06-11           5086.935   
4      AB-10015            3     2013-11-11            886.156   
..          ...          ...            ...                ...   
788    XP-21865           11     2014-11-18           2374.658   
789    YC-21895            5     2014-12-27           5454.350   
790    YS-21880            8     2014-12-22           6720.444   
791    ZC-21910           13     2014-12-06           8025.707   
792    ZD-21925            5     2014-12-06           1493.944   

     day_since_last_order  recency_score  
0                     184              2  
1                      19              5  
2                     259              2  
3                     203              2  
4                     415              1  
..                    ...            ...  
788                    43              4  
789                     4              5  
790                     9              5  
791                    25              5  
792                    25              5  

[793 rows x 6 columns]

In [ ]:
df_user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 793 entries, 0 to 792
Data columns (total 6 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Customer ID           793 non-null    object        
 1   order_count           793 non-null    int64         
 2   max_order_date        793 non-null    datetime64[ns]
 3   total_order_value     793 non-null    float64       
 4   day_since_last_order  793 non-null    int64         
 5   recency_score         793 non-null    int64         
dtypes: datetime64[ns](1), float64(1), int64(3), object(1)
memory usage: 37.3+ KB


Buat binning dari total transaksi (order) yang terdiri dari 5 bins dengan batas-batasnya merupakan min, P20, P40, P60, P80, max dan beri label 1 sampai 5 dari bin terendah ke tertinggi sebagai skor frequency

In [ ]:
df_user['frequency_score'] = pd.cut(df_user['order_count'],
                                    bins=[df_user['order_count'].min(),
                                          np.percentile(df_user['order_count'], 20),
                                          np.percentile(df_user['order_count'], 40),
                                          np.percentile(df_user['order_count'], 60),
                                          np.percentile(df_user['order_count'], 80),
                                          df_user['order_count'].max()],
                                    labels=[1, 2, 3, 4, 5],
                                    include_lowest=True).astype(int)
df_user

Customer ID  order_count max_order_date  total_order_value  \
0      AA-10315            5     2014-06-30           5563.560   
1      AA-10375            9     2014-12-12           1056.390   
2      AA-10480            4     2014-04-16           1790.512   
3      AA-10645            6     2014-06-11           5086.935   
4      AB-10015            3     2013-11-11            886.156   
..          ...          ...            ...                ...   
788    XP-21865           11     2014-11-18           2374.658   
789    YC-21895            5     2014-12-27           5454.350   
790    YS-21880            8     2014-12-22           6720.444   
791    ZC-21910           13     2014-12-06           8025.707   
792    ZD-21925            5     2014-12-06           1493.944   

     day_since_last_order  recency_score  frequency_score  
0                     184              2                2  
1                      19              5                5  
2                     259              2                1  
3                     203              2                3  
4                     415              1                1  
..                    ...            ...              ...  
788                    43              4                5  
789                     4              5                2  
790                     9              5                4  
791                    25              5                5  
792                    25              5                2  

[793 rows x 7 columns]

In [ ]:
df_user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 793 entries, 0 to 792
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Customer ID           793 non-null    object        
 1   order_count           793 non-null    int64         
 2   max_order_date        793 non-null    datetime64[ns]
 3   total_order_value     793 non-null    float64       
 4   day_since_last_order  793 non-null    int64         
 5   recency_score         793 non-null    int64         
 6   frequency_score       793 non-null    int64         
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 43.5+ KB


Buat binning dari total nilai order (order value) yang terdiri dari 5 bins dengan batas-batasnya merupakan min, P20, P40, P60, P80, max dan beri label 1 sampai 5 dari bin terendah ke tertinggi sebagai skor monetary

In [ ]:
df_user['monetary_score'] = pd.cut(df_user['total_order_value'],
                                    bins=[df_user['total_order_value'].min(),
                                          np.percentile(df_user['total_order_value'], 20),
                                          np.percentile(df_user['total_order_value'], 40),
                                          np.percentile(df_user['total_order_value'], 60),
                                          np.percentile(df_user['total_order_value'], 80),
                                          df_user['total_order_value'].max()],
                                    labels=[1, 2, 3, 4, 5],
                                    include_lowest=True).astype(int)
df_user

Customer ID  order_count max_order_date  total_order_value  \
0      AA-10315            5     2014-06-30           5563.560   
1      AA-10375            9     2014-12-12           1056.390   
2      AA-10480            4     2014-04-16           1790.512   
3      AA-10645            6     2014-06-11           5086.935   
4      AB-10015            3     2013-11-11            886.156   
..          ...          ...            ...                ...   
788    XP-21865           11     2014-11-18           2374.658   
789    YC-21895            5     2014-12-27           5454.350   
790    YS-21880            8     2014-12-22           6720.444   
791    ZC-21910           13     2014-12-06           8025.707   
792    ZD-21925            5     2014-12-06           1493.944   

     day_since_last_order  recency_score  frequency_score  monetary_score  
0                     184              2                2               5  
1                      19              5                5               2  
2                     259              2                1               3  
3                     203              2                3               5  
4                     415              1                1               1  
..                    ...            ...              ...             ...  
788                    43              4                5               3  
789                     4              5                2               5  
790                     9              5                4               5  
791                    25              5                5               5  
792                    25              5                2               2  

[793 rows x 8 columns]

In [ ]:
df_user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 793 entries, 0 to 792
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Customer ID           793 non-null    object        
 1   order_count           793 non-null    int64         
 2   max_order_date        793 non-null    datetime64[ns]
 3   total_order_value     793 non-null    float64       
 4   day_since_last_order  793 non-null    int64         
 5   recency_score         793 non-null    int64         
 6   frequency_score       793 non-null    int64         
 7   monetary_score        793 non-null    int64         
dtypes: datetime64[ns](1), float64(1), int64(5), object(1)
memory usage: 49.7+ KB


Buat kolom nama segmen berdasarkan skor recency dan frequency

In [ ]:
df_user['segment'] = np.select(
    [(df_user['recency_score']==5) & (df_user['frequency_score']>=4),
     (df_user['recency_score'].between(3,4)) & (df_user['frequency_score']>=4),
     (df_user['recency_score']>=4) & (df_user['frequency_score'].between(2,3)),
     (df_user['recency_score']<=2) & (df_user['frequency_score']==5),
     (df_user['recency_score']==3) & (df_user['frequency_score']==3),
     (df_user['recency_score']==5) & (df_user['frequency_score']==1),
     (df_user['recency_score']==4) & (df_user['frequency_score']==1),
     (df_user['recency_score']<=2) & (df_user['frequency_score'].between(3,4)),
     (df_user['recency_score']==3) & (df_user['frequency_score']<=2),
     (df_user['recency_score']<=2) & (df_user['frequency_score']<=2)],
    ['01-Champion', '02-Loyal Customers', '03-Potential Loyalists', "04-Can't Lose Them", '05-Need Attention',
     '06-New Customers', '07-Promising', '08-At Risk', '09-About to Sleep', '10-Hibernating']
)
df_user

Customer ID  order_count max_order_date  total_order_value  \
0      AA-10315            5     2014-06-30           5563.560   
1      AA-10375            9     2014-12-12           1056.390   
2      AA-10480            4     2014-04-16           1790.512   
3      AA-10645            6     2014-06-11           5086.935   
4      AB-10015            3     2013-11-11            886.156   
..          ...          ...            ...                ...   
788    XP-21865           11     2014-11-18           2374.658   
789    YC-21895            5     2014-12-27           5454.350   
790    YS-21880            8     2014-12-22           6720.444   
791    ZC-21910           13     2014-12-06           8025.707   
792    ZD-21925            5     2014-12-06           1493.944   

     day_since_last_order  recency_score  frequency_score  monetary_score  \
0                     184              2                2               5   
1                      19              5                5               2   
2                     259              2                1               3   
3                     203              2                3               5   
4                     415              1                1               1   
..                    ...            ...              ...             ...   
788                    43              4                5               3   
789                     4              5                2               5   
790                     9              5                4               5   
791                    25              5                5               5   
792                    25              5                2               2   

                    segment  
0            10-Hibernating  
1               01-Champion  
2            10-Hibernating  
3                08-At Risk  
4            10-Hibernating  
..                      ...  
788      02-Loyal Customers  
789  03-Potential Loyalists  
790             01-Champion  
791             01-Champion  
792  03-Potential Loyalists  

[793 rows x 9 columns]

In [ ]:
df_user.describe()

order_count  total_order_value  day_since_last_order  recency_score  \
count   793.000000         793.000000            793.000000     793.000000   
mean      6.316520        2896.848500            163.928121       3.003783   
std       2.550885        2628.670117            189.632742       1.419110   
min       1.000000           4.833000              0.000000       1.000000   
25%       5.000000        1146.050000             37.000000       2.000000   
50%       6.000000        2256.394000             93.000000       3.000000   
75%       8.000000        3785.276000            226.000000       4.000000   
max      17.000000       25043.050000           1166.000000       5.000000   

       frequency_score  monetary_score  
count       793.000000      793.000000  
mean          2.843632        3.000000  
std           1.426928        1.415998  
min           1.000000        1.000000  
25%           2.000000        2.000000  
50%           3.000000        3.000000  
75%           4.000000        4.000000  
max           5.000000        5.000000

Tampilkan summary dari RFM segmentation berupa banyaknya pengguna, rata-rata dan median dari total order, total order value, dan jumlah hari sejak order terakhir

In [ ]:
summary = pd.pivot_table(df_user, index='segment',
                         values=['Customer ID', 'day_since_last_order', 'order_count', 'total_order_value'],
                         aggfunc={'Customer ID': pd.Series.nunique,
                                  'day_since_last_order': [np.mean, np.median],
                                  'order_count': [np.mean, np.median],
                                  'total_order_value':[np.mean, np.median]})
summary['pct_unique'] = (summary['Customer ID'] / summary['Customer ID'].sum() * 100).round(1)
summary

Customer ID day_since_last_order        order_count  \
                           nunique                 mean median        mean   
segment                                                                      
01-Champion                     66            14.121212   13.5    9.621212   
02-Loyal Customers             108            67.768519   68.0    9.425926   
03-Potential Loyalists         148            29.736486   32.5    6.081081   
04-Can't Lose Them              37           240.783784  200.0    9.972973   
05-Need Attention               53            94.490566   96.0    6.509434   
06-New Customers                21            12.095238   10.0    3.428571   
07-Promising                    27            49.962963   49.0    3.333333   
08-At Risk                     100           253.470000  202.5    6.890000   
09-About to Sleep               53            96.245283   93.0    3.886792   
10-Hibernating                 180           396.527778  322.0    3.805556   

                              total_order_value            pct_unique  
                       median              mean     median             
segment                                                                
01-Champion               9.0       4180.286726  3734.9705        8.3  
02-Loyal Customers        9.0       3935.686984  3252.6430       13.6  
03-Potential Loyalists    6.0       3022.441259  2382.7365       18.7  
04-Can't Lose Them        9.0       4333.072089  3133.9240        4.7  
05-Need Attention         7.0       2937.473168  2533.1640        6.7  
06-New Customers          4.0       1245.132800   883.4090        2.6  
07-Promising              4.0       2092.656844  1205.5840        3.4  
08-At Risk                7.0       3283.584423  2784.4740       12.6  
09-About to Sleep         4.0       1952.874845  1058.1080        6.7  
10-Hibernating            4.0       1768.924211  1147.4150       22.7

1. Pengguna paling banyak berada pada segmen Hibernating (180 atau 22.7%), Potential Loyalist (148 atau 18.7%), dan Loyal Customer (108 atau 13.6%).
2.  Program khusus yang berfokus pada urgensi bertransaksi untuk Loyal Customer (108 atau 13.6%) dapat dibuat untuk membuat mereka bertransaksi kembali dalam waktu dekat sehingga naik ke segment Champion
3. Program khusus yang berfokus pada jumlah untuk potential Loyalist (!48 atau 18.7%) dapat dibuar untuk membuat mereka lebih sering bertransaksi sehingga bisa naik ke segmen Champion.
4. Program khusus untuk Hibernating (180 atau 22.7%) dapat dibuat untuk membuat mereka kembali bertrankasi walaupun belum begitu sering sehingga bisa naik ke segment New Customer atau bahkan Potential Loyalist.